azext_edge/edge/providers/adr/validator.py

```diff
@@ -0,0 +1,636 @@
+
+        # Use local bundled schema for OPC UA
+        if self.endpoint_type.lower() in ["microsoft.opcua", "microsoft.deviceregistry.opcua", "opcua"]:
+            metadata = self._load_local_opcua_metadata()
+            if metadata:
+                self._METADATA_CACHE[cache_key] = metadata
+                return metadata
+            # Fall through to OCI fetch as fallback
+
+        try:
+            # Step 1: Get IoT Operations management client
+            from ...vendor.clients.iotopsmgmt import MicrosoftIoTOperationsManagementService
+
+            iotops_client: MicrosoftIoTOperationsManagementService = get_iotops_mgmt_client(
+                subscription_id=self.cmd.cli_ctx.data.get("subscription_id"),
+                endpoint=self.cmd.cli_ctx.cloud.endpoints.resource_manager,
+            )
+
+            # Step 2: List all connector templates in the instance
+            connector_templates = list(
+                iotops_client.akri_connector_template.list_by_instance_resource(
+                    resource_group_name=self.resource_group_name, instance_name=self.instance_name
+                )
+            )
+
+            # Step 3: Find matching connector template
+            matched_template = None
+            for template in connector_templates:
+                template_name = template.get("name")
+                device_endpoint_types = template.get("properties", {}).get("deviceInboundEndpointTypes", [])
+
+                for endpoint_type_info in device_endpoint_types:
+                    et = endpoint_type_info.get("endpointType")
+                    ev = endpoint_type_info.get("version")
+
+                    # Match endpoint type
+                    if not et or et.lower() != self.endpoint_type.lower():
+                        continue
+
+                    # Match version (if both specified, they must match; if either is None, match)
+                    if self.endpoint_version and ev and self.endpoint_version != ev:
+                        continue
+
+                    logger.info(
+                        f"Matched connector template '{template_name}' for endpoint type "
+                        f"'{self.endpoint_type}' version '{self.endpoint_version or ev}'"
+                    )
+                    matched_template = template
+                    break
+
+                if matched_template:
+                    break
+
+            if not matched_template:
+                logger.warning(
+                    f"No connector template found for endpoint type '{self.endpoint_type}' "
+                    f"version '{self.endpoint_version}'. Validation will be skipped."
+                )
+                return {}
+
+            # Step 4: Extract connectorMetadataRef
+            connector_metadata_ref = matched_template.get("properties", {}).get("connectorMetadataRef")
+
+            if not connector_metadata_ref:
+                logger.warning(
```

Comment:
General comment to change the behavior to throw if we decide that a matching connector must be present. All connectors are required to have a metadata ref.